core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -87,6 +87,12 @@ class LogSegment private[log] (val log: FileRecords,
       // we will recover the segments above the recovery point in recoverLog()
       // in any case so sanity checking them here is redundant.
       txnIndex.sanityCheck()
+      // Failing to sanity check the timeIndex can result in a scenario where log segments are
+      // prematurely deleted (before breaching retention periods) if the index file was not resized
+      // to disk successfully.
+      // KAFKA-10207
+      timeIndex.sanityCheck()
+      offsetIndex.sanityCheck()
```

Comment:
As you say, these were originally removed for performance reasons. We'd want to verify the performance impact of adding them back. cc @junrao @efeg

Reply:
If the concern here is startup performance, we could instead fallback to using the mtime of the file when the last offset and timestamp are both 0.

Reply:
Did we consider performing the sanity check lazily, eg. when the index is loaded for the first time?

Reply:
@Johnny-Malizia : Thanks for reporting this. A couple of comments on this.

(1) It seems that there are cases where the index file is not trimmed properly. Every rolled segment calls LogSegment.onBecomeInactiveSegment(), which does trimming. So, we probably should just fix the trim logic there.

(2) The intention of KIP-263 is to avoid unnecessary loading (and the sanity check) of indexes never used. This is mostly achieved through lazy indexes (actual index file only opened lazily on first use). Since the biggest cost of loading an index file is probably the I/O part, and the sanity check is computational only and relatively cheap, we can probably just do the index sanity check on index opening. This way, in the common case, we only pay the index sanity checking cost on active segments.

Reply:
Thank you for the feedback here. 

While I agree that the trimming logic should be corrected if possible, it seems like an issue like this is out of our hands to some degree. The current logic seems to be sound and works on a newer version of the jvm. The issue was related to reducing the mmaped file's length not actually happening and failing silently.  I'm open to suggestions to work around this, but even working around the issue for this specific jvm and this specific version of zfs it seems plausible this same issue could crop up again in another jvm or another storage driver whereas checking at read time catches this particularly nasty issue.

I think that a good compromise here would be to follow through as you and @dhruvilshah3 have suggested and apply the sanity check when the index is first loaded as the check itself is cheap enough and we would avoid loading *every* segment at startup.